Implement Go to Type Definition

jez opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->


### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

=> #1500

### Commit summary

I recommend reviewing by commit.

- **Implement test harness for TypeDefAssertion** (2b98bac6c)

  It ended up growing support to assert that one 
  textDocument/typeDefinition request respond with multiple locs (for 
  cases like to go type definition on a union type).

- **Boilerplate for new TypeDefinition handler** (cd2ba2b42)


- **Implement handleTextDocumentTypeDefinition** (63b9b936e)

  Note that we never try to add a loc for untyped, because it wiil try to 
  jump to T::Utils::RuntimeProfiled.

- **Add some tests** (fafab8684)

  I did some testing locally to check that adding deleting things that I 
  would expect to make this test fail actually made it fail.




### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

jez commented:
cc @jvilk-stripe still wip but just in case you're curious

I'll probably ping you tomorrow to see how far we should go to implement this. The bare essentials work right now.

jvilk-stripe commented on main/lsp/requests/type_definition.cc:
```diff
@@ -0,0 +1,76 @@
+                    addLocIfExists(*gs, result, loc);
+                }
+            } else if (fileIsTyped && resp->isDefinition()) {
+                // TODO(jez) Not clear what go to type def on a definition should do...
```
TypeScript just brings you to the definition site, so I think the go-to-def behavior works here too.

jvilk-stripe commented on main/lsp/requests/type_definition.cc:
```diff
@@ -0,0 +1,72 @@
+                }
+            } else if (fileIsTyped && resp->isSend()) {
+                auto sendResp = resp->isSend();
+                for (auto loc : locsForType(*gs, sendResp->dispatchResult->returnType)) {
```
Hmm, is this the expected semantics? I'd expect go-to-type-definition on a method to go to the method, which is what typescript does. It's somewhat surprising that it jumps to the definition site of the return value.

jez replied:
Oh interesting.

It seems that if you're on a send, "Go to Definition" should go to the method's definition, and "Go to Type Definition" should then go to the definition of the type that method returns. I kind of seeing it as skipping a jump: rather than having to "Go to Definition", find the sig, and then "Go to Definition" on the `returns(...)` within the sig, "Go to Type Definition" goes there in one jump.

I kind of built this feature w.r.t. what features I wished i had when using e.g. clangd, rather than looking at what other LSP clients did.

jvilk-stripe commented:
I'm happy to back down if you have a strong opinion. The rest of the PR is super great!

jez commented:
> * Go to Type Definition in `# typed: false` (doesn't work for certain entities, but works on others).

Good suggestion, added a test. It uncovered a bug in DefLocSaver, fixed in 564cf3b.

> * What's an `AppliedType`? Is that typecase exercised?

That's a generic type. I forgot to add tests for generics. I added some.

Unrelated: one thing which i couldn't write a test for was this:

```ruby
# typed: true

class A; end
#     ^ type-def: A

class TestClass
  a = T.let(nil, T.nilable(A))
  puts a
  #    ^ type: A
end
```

because the definition for `NilClass` is in an RBI.

I can sort of fake it: if I add

```ruby
class NilClass; end
```

to the top of the file, it will register an addition loc to the definition of `NilClass`, and since this loc is not in an RBI, it takes precedence as the "canonical" loc:

https://github.com/sorbet/sorbet/blob/6b3d747e33215f093ec50e0d1a6d82486dab2f62/core/Symbols.cc#L1144-L1148

but that seems like a hack, and that's not going to be what the editor does.

How do you test jump to definition for things defined in the standard library?
